app/forms/analyses/gwas/form.rb

```diff
@@ -0,0 +1,94 @@
+module Analyses
+  module Gwas
+    class Form < BaseForm
+      include Reform::Form::ActiveModel
+
+      model :analysis
+
+      property :analysis_type
+      property :name
+      property :genotype_data_file_id
+      property :map_data_file_id
+      property :phenotype_data_file_id
+
+      validates :name, presence: true
+      validates :genotype_data_file, presence: true
+      validates :phenotype_data_file, presence: true
+
+      validate do
+        geno = parse_data_file(genotype_data_file, genotype_data_parser)
+        map = parse_data_file(map_data_file, map_data_parser)
+        pheno = parse_data_file(phenotype_data_file, phenotype_data_parser)
+
+        if geno && pheno
```

Comment:
It's up to you. I think separate checks would allow for shorter code that is easier to read, in this case. I may be wrong.